design/StableFunctions.md

```diff
@@ -0,0 +1,126 @@
+# Stable Functions, Stable Objects, and Stable Classes
+
+Functions, objects, and classes can also be persisted with enhanced orthogonal persistence, i.e. they been used in Motoko's stable variables and persistent actor variables.
+
+Note: This is not supported with classical persistence.
+
+With this, the only non-persistent (non-stable) values remain to be lambdas, async handles (aka continuations, futures), and local async function references.
+
+## Example
+
+With this, the following program can now be used for orthogonal persistence. For this example, you need a [specific base library branch](https://github.com/dfinity/motoko-base/tree/luc%2Fstable-functions), which contains slight modifications for supporting stable functions.
+
+```
+import HashMap "base/HashMap";
+import Hash "base/Hash";
+import Nat "base/Nat";
+import Iter "base/Iter";
+
+persistent actor {
+    let hashMap = HashMap.HashMap<Nat, Text>(10, Nat.equal, Hash.hash); // finally persistent!
+
+    public func populate(): async() {
+        hashMap.put(1, "A");
+        hashMap.put(2, "B");
+        hashMap.put(3, "C");
+    };
+
+    public func list(): async [(Nat, Text)] {
+        Iter.toArray<(Nat, Text)>(hashMap.entries());
+    };
+}
+```
+
+## Stable Functions and Stable Scopes
+
+A stable function is a named non-async local function in a stable scope, only closing over variables of a stable type.
+
+A stable scope is:
+* the main actor,
+* an actor class,
+* a module imported with a unique identifier from a stable scope,
+* a named non-async function in a stable scope,
+* a class in a stable scope, or,
+* a named object in a stable scope.
+
+Generic type parameters of stable functions and stable classes are bounded to stable types.
```

Comment:
```suggestion
Generic type parameters of stable functions and stable classes are bounded to stable types.
```
I'm not sure this is fully implemented given the counter-example in #5143. Maybe we need another sort of bound, beyond the current Scope|Type - Scope|Type|StableType. Then we can require type parameters of stable functions to be StableType, and type parameters of local functions to be either Type Or StableType. 
Parameters with bound StableType can only be instantiated at stable types, and only parameters with bound StableType are stable types (currently any parameter is a stable type).